semgrep-core/src/parsing/tree_sitter/Parse_swift_tree_sitter.ml

```diff
@@ -1240,40 +1248,48 @@ and map_expression (env : env) (x : CST.expression) : G.expr =
 
 and map_for_statement (env : env)
     ((v1, v2, v3, v4, v5, v6, v7, v8, v9) : CST.for_statement) =
-  let v1 = (* "for" *) token env v1 in
-  let v2 =
-    match v2 with
-    | Some x -> map_try_operator env x
-    | None -> todo env ()
-  in
-  let v3 =
-    match v3 with
-    | Some tok -> (* "await" *) token env tok
-    | None -> todo env ()
-  in
-  let v4 = map_binding_pattern_no_expr env v4 in
-  let v5 =
-    match v5 with
-    | Some x -> map_type_annotation env x
-    | None -> todo env ()
+  let for_tok = (* "for" *) token env v1 in
+  let pat =
+    let pat_init = map_binding_pattern_no_expr env v4 in
+    let add_pat_ty pat =
+      match v5 with
+      | Some x -> G.PatTyped (pat, map_type_annotation env x)
+      | None -> pat
+    in
+    let add_pat_where pat =
+      match v8 with
+      | Some x -> G.PatWhen (pat, map_where_clause env x)
+      | None -> pat
+    in
+    let add_pat_try pat =
+      match v2 with
+      | Some x (* "try" *) ->
+          G.OtherPat (("Try", map_try_operator env x), [ G.P pat ])
+      | None -> pat
+    in
+    let add_pat_await pat =
+      match v3 with
+      | Some tok (* "await" *) ->
+          G.OtherPat (("Await", token env tok), [ G.P pat ])
+      | None -> pat
+    in
+    pat_init |> add_pat_ty |> add_pat_where |> add_pat_try |> add_pat_await
```

Comment:
I think this is a good candidate for simplification via `let*` (as syntactic sugar for `Option.bind`) but I'm also fine with it as-is.